### PR TITLE
July 21 Updates and Questions!

### PR DESCRIPTION
**Pull request by AinaMerch**
I have marked the places each question refers to in the code as well.

Question 1:
I'm trying to set the synchronization property of node mappings but I can somehow set the synchronization lock but not get it from the original? Do I type in a string, or does it not need to be customized at all? (Also see issue 8)

Question 2:
For the edge mappings there are 2 properties called source mappings and target mappings that I can get but not set. I am unsure whether the source and target expressions set those automatically which is why I can't set them or there's something else. (Also see issue 8)

Question 3:
I know you said I shouldn't use the copy trick for other elements but since this is style for edges which should also be a self contained leaf element it works here right?

Question 4:
I just want to know if I'm on the right path with the way I'm going about trying to create the tools. I can use instanceof here yes? Since the type of the owned tool would be node creation or edge creation or something like that? 

Question 5:
It was giving me red squigglies without the casting but it worked without casting for the node mapping above?

Question 6:
Can I use copy for tools or not? Why would that not work if not?

Question 7:
For the internal action sub node I use the same helper function createProductLineNodeMapping that uses the copy for copying style. The internal action also has a gradient though. Does that get copied too or would I need to do something else? (Also see issue 8)

Issue 8:
I tried to actually view the conversion in the run time configuration mode so that I could see whether or not what I was trying to do was actually happening and what was or wasn't showing up/getting translated correctly. But when I tried running operator the way you showed me on the test file you'd provided me with way back when to actually translate a statemachine to productline I got the following error.
Screenshot from 2024-07-15 03-09-41

### CODE REVIEW

**Comment by adisandro** [comment]
> Question 1: I'm trying to set the synchronization property of node mappings but I can somehow set the synchronization lock but not get it from the original? Do I type in a string, or does it not need to be customized at all? (Also see issue 8)

You can read boolean properties with the `isX()` api, rather than `getX()`.
In this case `isSynchronizationLock()`.

> Question 2: For the edge mappings there are 2 properties called source mappings and target mappings that I can get but not set. I am unsure whether the source and target expressions set those automatically which is why I can't set them or there's something else. (Also see issue 8)

It's a list of mappings, if you check the return type of `getSourceMapping()` it's `EList` (which is the EMF equivalent of a plain java List).
You have to loop through the original ones, and add new ones like you'd add to a list, `edgeMapping.getSourceMapping().add(newSource)`.

> Question 3: I know you said I shouldn't use the copy trick for other elements but since this is style for edges which should also be a self contained leaf element it works here right?

Yes it should :)

> Question 4: I just want to know if I'm on the right path with the way I'm going about trying to create the tools. I can use instanceof here yes? Since the type of the owned tool would be node creation or edge creation or something like that?

Yes, you can totally use it and choose different actions based on that, nothing wrong about it.

> Question 5: It was giving me red squigglies without the casting but it worked without casting for the node mapping above?

No that's wrong, it's different for edges.
The label expression is attached to the label style (center, begin or end), not the edge style.
The code is going to be a bit more complicated than that.
(and you can't copy the expression as-is, that's one of the things that changes on the PL side)

> Question 6: Can I use copy for tools or not? Why would that not work if not?

You can't just copy when you have references to things you are also recreating.
A tool has node mappings in this case, and they have to be the new node mappings you have created before.
You had `InitialState`, you have created `PLInitialState`, then the tool that creates initial states now has to create PL initial states.

> Question 7: For the internal action sub node I use the same helper function createProductLineNodeMapping that uses the copy for copying style. The internal action also has a gradient though. Does that get copied too or would I need to do something else? (Also see issue 8)

It should be copied, but again if you launch it in debug mode you can check step by step what happens.

> Issue 8: I tried to actually view the conversion in the run time configuration mode so that I could see whether or not what I was trying to do was actually happening and what was or wasn't showing up/getting translated correctly. But when I tried running operator the way you showed me on the test file you'd provided me with way back when to actually translate a statemachine to productline I got the following error. Screenshot from 2024-07-15 03-09-41

That's a runtime error in your code.
If you Debug as instead of Run as when you launch the runtime Eclipse instance, you can set breakpoints in your code and inspect variables step by step to see what happens.

In any case, it's at line 110 where you force a cast. It compiles but that doesn't mean it's going to be correct at runtime when you deal with the actual objects.
You have a `DiagramExtensionDescription` which does not have a default layer, you cast it to make it have it but at runtime you'll get an object that indeed does not have it, and it tells you so. It has just `getLayers()` and that's fine, they are two different objects. The rational is that the PL diagram is an extension of an existing diagram, which already has its default layer, so you're adding an additional layer instead.

**Comment by adisandro** [comment]
> JULY 21: Issue 1: Operator ToProductLine execution error The object 'org.eclipse.sirius.diagram.model.business.internal.description.spec.NodeMappingSpec PLState' is not contained in a resource.
> 
> When I try to run the class I get this error and upon going through with the debugger I saw that it stops at the place where I marked as issue 1 but I didn't even write that code so I'm unsure of how to fix the issue.

Yeah it happens there because it's where the output model gets saved to a file.
The real error is that object not being assigned to a container, it's dangling.

You are creating new node mappings for source and target when you create edge mappings, instead of reusing the ones you have already created earlier.
Apart from being wrong conceptually, it also means these new mappings are not inserted in their proper container layer `newLayer.getNodeMappings().add(..)`, hence the error message.

Each object in EMF must be contained by something, except the root. Each reference X, that gets translated into the API `setX` (multiplicity 1) or `getX().add()` (multiplicity many), has a setting `Container: true/false` in the metamodel. Source and target in the edge mappings are not containers.

> Question 2: Is this the right way to set the DirectEditLabel?

Ouch, no, sorry that's my fault.
Ignore that label, it should be imported from the base PL diagram that you don't have access to.
I too don't know yet how to do it.

> Question 3: Just want to know if I'm on the right track with the way I'm mapping

Yes makes sense, although I would make the value type more specific if possible. Do node and edge mappings share a common super class by any chance, or is it really Object?
You can reuse the map for the source/target node mappings (see question 1).

> Question 4/5: Can I access the original Mappings like that? If so, do I use getLabel or getName for the original Mapping?

`getName` is required to be unique, so that one (also to be used when you put them in the map).

> Question 6: Alternate way of accessing mapping but this may not take into account that it may have more than one mapping.

You can't simply get the name of the original tool, you have to get the name of the original mapping attached to the original tool :)
I follow a convention to name them similarly, but you can't assume that to always be the case.

> Question 7: How do you go about taking this from statemachine to convert? Also how do I create a FIrstModalOperation because it says not defined.

No, here you don't convert, you just create the new one.
There's an external api that takes it into account, it would be too complex to put it here directly.

`vToolFactory.createChangeContext()` is the specific operation you need, then you assign it to the `setFirstModelOperation()`.

File: plugins/Core-Extensions/ProductLine/edu.toronto.cs.se.mmint.productline.design/src/edu/toronto/cs/se/mmint/productline/design/ToProductLine.java [hunk 1]
```diff
@@ -250,62 +243,148 @@ private ContainerMapping createProductLineContainerMapping(ContainerMapping orig
...
 	    	NodeMapping newMapping = (NodeMapping) map.get(originalMapping.getName());
 	    	createTool.getNodeMappings().add(newMapping);
 	    }
-	    
-	    createTool.setIconPath("/edu.toronto.cs.se.modelepedia.statemachine.edit/icons/full/obj16/"+newString+".gif");
-//	    originalCreateTool.getInitialOperation().getFirstModelOperations();
 	    /*
-	     * QUESTION 7
+	     * QUESTION 1: How do I add these to the actual creation tool? Are these the right elements in the first
+	     * place?
```
Comment (adisandro) [inline plugins/Core-Extensions/ProductLine/edu.toronto.cs.se.mmint.productline.design/src/edu/toronto/cs/se/mmint/productline/design/ToProductLine.java:257]: I'm sorry, hard to say without a computer.
When you click on them in the diagram spec, the properties view should show what type of variable they are (or opening the diagram spec as a text file, their xsi:type field will show it).

File: plugins/Core-Extensions/ProductLine/edu.toronto.cs.se.mmint.productline.design/src/edu/toronto/cs/se/mmint/productline/design/ToProductLine.java [hunk 2]
```diff
@@ -250,62 +243,148 @@ private ContainerMapping createProductLineContainerMapping(ContainerMapping orig
...
+//        createTool.getVariable().;
+        
+        /*
+         * QUESTION 2: We didn't discuss this in the meeting but I'm assuming the initial operation does need
+         * to be created before we can get it?
+         */
+	    var initialOperation = this.out.vToolFactory.createInitialOperation();
+	    createTool.setInitialOperation((InitialNodeCreationOperation) initialOperation);
+	    /*
+	     * QUESTION 3: Just checking the approach?
```
Comment (adisandro) [inline plugins/Core-Extensions/ProductLine/edu.toronto.cs.se.mmint.productline.design/src/edu/toronto/cs/se/mmint/productline/design/ToProductLine.java:274]: Looks good.

File: plugins/Core-Extensions/ProductLine/edu.toronto.cs.se.mmint.productline.design/src/edu/toronto/cs/se/mmint/productline/design/ToProductLine.java [hunk 3]
```diff
@@ -250,62 +243,148 @@ private ContainerMapping createProductLineContainerMapping(ContainerMapping orig
...
+	    // or do we use createSelectContainerVariable()? or createElementVariable()?
+	    var containerView = this.out.vToolFactory.createContainerViewVariable();
+	     
+        container.setName("container");
+        containerView.setName("containerView");
+//        createTool.getVariable().;
+        
+        /*
+         * QUESTION 2: We didn't discuss this in the meeting but I'm assuming the initial operation does need
+         * to be created before we can get it?
```
Comment (adisandro) [inline plugins/Core-Extensions/ProductLine/edu.toronto.cs.se.mmint.productline.design/src/edu/toronto/cs/se/mmint/productline/design/ToProductLine.java:269]: We probably missed this, not sure without a computer.
The fact that you have to explicitly cast it though is not a good sign, it should just be assignable as-is.

File: plugins/Core-Extensions/ProductLine/edu.toronto.cs.se.mmint.productline.design/src/edu/toronto/cs/se/mmint/productline/design/ToProductLine.java [hunk 4]
```diff
@@ -250,62 +243,148 @@ private ContainerMapping createProductLineContainerMapping(ContainerMapping orig
...
+	    
+	    for (var originalMapping : originalCreateTool.getEdgeMappings()) {
+	    	EdgeMapping newMapping = (EdgeMapping) map.get(originalMapping.getName());
+	    	createTool.getEdgeMappings().add(newMapping);
+	    }
 	    /*
-	     * QUESTION 6
+	     * QUESTION 4:
+	     * Are source and target variables and source and target view variables the same type as above?
+	     * If not what are they?
```
Comment (adisandro) [inline plugins/Core-Extensions/ProductLine/edu.toronto.cs.se.mmint.productline.design/src/edu/toronto/cs/se/mmint/productline/design/ToProductLine.java:309]: They're the same, but unfortunately I'm not sure what they should be above.

File: plugins/Core-Extensions/ProductLine/edu.toronto.cs.se.mmint.productline.design/src/edu/toronto/cs/se/mmint/productline/design/ToProductLine.java (hunk 1 above)
Comment (AinaMerch) [reply plugins/Core-Extensions/ProductLine/edu.toronto.cs.se.mmint.productline.design/src/edu/toronto/cs/se/mmint/productline/design/ToProductLine.java:257]: <variable name="container"/>
<viewVariable name="containerView"/>

Yeah I already checked that. This is what it says.
In the drop down I have createNameVariable(), createContainerViewVariable(), createSelectContainerVariable(), createAcceleoVariable(), createElementVariable(), createElementSelectVariable(), createElementDropVariable(), createElementViewVariable(). There is no createVariable(). 

image
This is what the properties view looks like too. Hence, the confusion.

Comment (AinaMerch) [reply plugins/Core-Extensions/ProductLine/edu.toronto.cs.se.mmint.productline.design/src/edu/toronto/cs/se/mmint/productline/design/ToProductLine.java:257]: Also, when I try using createTool.setVariable()
The method setVariable(NodeCreationVariable) in the type NodeCreationDescription is not applicable for the arguments (NameVariable)

I get this but I can't seem to be able to create a NodeCreationVariable

Comment (adisandro) [reply plugins/Core-Extensions/ProductLine/edu.toronto.cs.se.mmint.productline.design/src/edu/toronto/cs/se/mmint/productline/design/ToProductLine.java:257]: Yep, my best guess is that a NodeCreationVariable can be created from another factory.
What about the dToolFactory?

**Comment by adisandro** [comment]
I don't see any other question in the code other than the factory issue. Does it mean it is ready to be merged when you have fixed that one?
There are no compilation errors, but please make sure that it is runnable in the runtime instance before asking me to merge.

**Comment by AinaMerch** [comment]
> I don't see any other question in the code other than the factory issue. Does it mean it is ready to be merged when you have fixed that one? There are no compilation errors, but please make sure that it is runnable in the runtime instance before asking me to merge.

**Comment by AinaMerch** [comment]
Hi Alessio, it isn't in dToolFactory either, I'd checked that before. I had a few other bugs but they seem mostly sorted. I just need that piece.

**Comment by adisandro** [comment]
I found it in `dToolFactory.createNodeCreationVariable()`.

**Pull request merged**